release/pkg/tasks/hashrelease.go

```diff
@@ -0,0 +1,174 @@
+	r.exists = exists
+	r.err = err
+	ch <- r
+}
+
+// HashreleaseValidate validates the images in the hashrelease.
+// These images are checked to ensure they exist in the registry
+// as they should have been pushed in the standard build process.
+func HashreleaseValidate(cfg *config.Config) {
+	// TODO: Consider checking if hashreleae already exists first (NEW!)
+	images, err := hashrelease.RetrieveComponentsToValidate(cfg.OutputDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to get pinned version")
+	}
+	results := make(map[string]imageExistsResult, len(images))
+
+	ch := make(chan imageExistsResult)
+	for name, component := range images {
+		go imgExists(name, component, ch)
+	}
+	for range images {
+		res := <-ch
+		results[res.name] = res
+	}
+	failedImages := []string{}
+	for name, r := range results {
+		logrus.WithFields(logrus.Fields{
+			"image":  r.image,
+			"exists": r.exists,
+		}).Info("Image validation")
+		if r.err != nil || !r.exists {
+			logrus.WithError(r.err).WithField("image", name).Error("Error checking image")
+			failedImages = append(failedImages, name)
+		} else {
+			logrus.WithField("image", name).Info("Image exists")
+		}
+	}
+	failedCount := len(failedImages)
+	if failedCount > 0 {
+		logrus.WithField("images", strings.Join(failedImages, ", ")).
+			Fatalf("Failed to validate %d images, see above for details", failedCount)
+	}
+}
+
+// HashreleaseValidate publishes the hashrelease
+func HashreleasePush(cfg *config.Config) {
+	outputDir := cfg.OutputDir
+	sshConfig := command.NewSSHConfig(cfg.DocsHost, cfg.DocsUser, cfg.DocsKey, cfg.DocsPort)
+	name, err := hashrelease.RetrieveReleaseName(outputDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to get release name")
+	}
+	note, err := hashrelease.RetrievePinnedVersionNote(outputDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to get pinned version note")
+	}
+	calicoVersion, err := hashrelease.RetrievePinnedCalicoVersion(outputDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to get candidate name")
+	}
+	// TODO: send image to image scan server
+	releaseHash, err := hashrelease.RetrievePinnedVersionHash(outputDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to get release hash")
+	}
+	releaseVersion := version.Version(calicoVersion)
+	logrus.WithField("note", note).Info("Publishing hashrelease")
+	if err := hashrelease.PublishHashrelease(name, releaseHash, note, releaseVersion.Stream(), cfg.HashreleaseDir(), sshConfig); err != nil {
+		logrus.WithError(err).Fatal("Failed to publish hashrelease")
+	}
+}
+
+// HashreleaseCleanRemote cleans up old hashreleases on the docs host
+func HashreleaseCleanRemote(cfg *config.Config) {
+	sshConfig := command.NewSSHConfig(cfg.DocsHost, cfg.DocsUser, cfg.DocsKey, cfg.DocsPort)
+	logrus.Info("Cleaning up old hashreleases")
+	if err := hashrelease.DeleteOldHashreleases(sshConfig, -1); err != nil {
```

Comment:
Why do we pass -1 here just to have it defaulted to 400? 

This probably falls into the category of "Dead code making this more complicated than it needs to be".

We should either:
- Remove this argument from DeleteoldHashreleases, and always default to 400
- Remove the defaulting and require the caller pass in a value, and then pass in 400 here.